infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java

```diff
@@ -0,0 +1,171 @@
+
+  @Inject
+  public AsyncStoragePodWatcher(
+      KubernetesClientFactory kubernetesClientFactory,
+      UserManager userManager,
+      PreferenceManager preferenceManager,
+      WorkspaceRuntimes runtimes,
+      @Named("che.infra.kubernetes.async.storage.shutdown_timeout_min") long shutdownTimeoutMin,
+      @Named("che.infra.kubernetes.pvc.strategy") String pvcStrategy,
+      @Named("che.infra.kubernetes.namespace.allow_user_defined")
+          boolean allowUserDefinedNamespaces,
+      @Nullable @Named("che.infra.kubernetes.namespace.default") String defaultNamespaceName,
+      @Named("che.limits.user.workspaces.run.count") int runtimesPerUser) {
+    this.kubernetesClientFactory = kubernetesClientFactory;
+    this.userManager = userManager;
+    this.preferenceManager = preferenceManager;
+    this.runtimes = runtimes;
+    this.shutdownTimeoutSec = TimeUnit.MINUTES.toSeconds(shutdownTimeoutMin);
+
+    isAsyncStoragePodCanBeRun =
+        isAsyncStoragePodCanBeRun(
+            pvcStrategy, allowUserDefinedNamespaces, defaultNamespaceName, runtimesPerUser);
+  }
+
+  /**
+   * Checking current system configuration on ability to run Async Storage Pod. Will be checked next
+   * value of properties:
+   *
+   * <ul>
+   *   <li>che.infra.kubernetes.namespace.default=<username>-che
+   *   <li>che.infra.kubernetes.namespace.allow_user_defined=false
+   *   <li>che.infra.kubernetes.pvc.strategy=common
+   *   <li>che.limits.user.workspaces.run.count=1
+   * </ul>
+   */
+  private boolean isAsyncStoragePodCanBeRun(
+      String pvcStrategy,
+      boolean allowUserDefinedNamespaces,
+      String defaultNamespaceName,
+      int runtimesPerUser) {
+    return !allowUserDefinedNamespaces
+        && COMMON_STRATEGY.equals(pvcStrategy)
+        && runtimesPerUser == 1
+        && !isNullOrEmpty(defaultNamespaceName)
+        && defaultNamespaceName.contains("<username>");
+  }
+
+  @ScheduleDelay(
+      unit = TimeUnit.MINUTES,
+      initialDelay = 1,
+      delayParameterName = "che.infra.kubernetes.async.storage.shutdown_check_period_min")
+  public void check() {
+    if (isAsyncStoragePodCanBeRun) { // if system not support async storage mode do nothing
+      for (User user :
+          iterateLazily((maxItems, skipCount) -> userManager.getAll(maxItems, skipCount))) {
+        try {
+          String owner = user.getId();
+          Map<String, String> preferences = preferenceManager.find(owner);
+          String lastTimeAccess = preferences.get(LAST_ACTIVITY_TIME);
+          String namespace = preferences.get(LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE);
+          if (isNullOrEmpty(namespace)
+              || isNullOrEmpty(lastTimeAccess)
+              || isAnyRuntimeInProgress(owner)) {
+            continue;
+          }
+          long lastTimeAccessSec = parseLong(lastTimeAccess);
+          long epochSec = clock.instant().getEpochSecond();
+          if (epochSec - lastTimeAccessSec >= shutdownTimeoutSec) {
```

Comment:
`ZonedDateTime` has quite nice API to calculate with times, `isAfter`, `isBefore` etc. Please use it instead of pure longs.

Reply:
sorry, but don't get that the problem here

Reply:
I would prefer to use well defined and tested API over calculating longs, which is prone to errors.